### PR TITLE
Fixing problem with template to allow for deployment to openshift

### LINKED ISSUE #117: OpenShift template is broken
**Describe the bug**
When attempting to install the openshift template for deploying log anomaly detector it is broken.

https://raw.githubusercontent.com/AICoE/log-anomaly-detector/master/openshift/log-anomaly-detector.app.yaml

**To Reproduce**

**Expected behavior**

**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by zmhassan**
# PR Details

Currently the template doesn't work and is broken. We would need to make it work to proceed with deploying onto openshift.

## Description

The openshift instructions do not work and the deployment fails. We would need to have a working template to deploy LAD into paas

## Related Issue

Fixes #117 

## Motivation and Context

Would like to be able to deploy LAD via s2i with template.

## How Has This Been Tested
```bash
oc create official-template.yaml
```

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by zmhassan** [comment]
Do not merge yet. As this is work in progress.

File: openshift/official_template.yaml [hunk 1]
```diff
@@ -0,0 +1,162 @@
+apiVersion: template.openshift.io/v1
+kind: Template
+metadata:
+  name: log-anomaly-detector
+  namespace: aiops-prod-log-analysis
+objects:
+- apiVersion: v1
```
Comment (goern) [inline openshift/official_template.yaml]: indent level needs to be +2

File: openshift/official_template.yaml [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  kind: PersistentVolumeClaim
+  metadata:
+    labels:
+      app: log-anomaly-detector-f20
+    name: log-anomaly-detector-f20
+  spec:
+    accessModes:
+    - ReadWriteOnce
```
Comment (goern) [inline openshift/official_template.yaml]: indent level needs to be +2

File: openshift/official_template.yaml [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+      type: Git
+    strategy:
+      sourceStrategy:
+        from:
+          kind: ImageStreamTag
+          name: python:3.6
+          namespace: openshift
+      type: Source
+    triggers:
+    - imageChange: {}
```
Comment (goern) [inline openshift/official_template.yaml]: indent level needs to be +2

you should think about installing some plugin to your IDE that does autoformatting in a correct way.

File: openshift/official_template.yaml [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+      type: ImageChange
+    - type: ConfigChange
+- apiVersion: apps.openshift.io/v1
+  kind: DeploymentConfig
+  metadata:
+    labels:
+      app: log-anomaly-detector-f20
```
Comment (vpavlin) [inline openshift/official_template.yaml]: Why does deployment config name mention `f20`? What does it mean?

File: openshift/official_template.yaml [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+        dnsPolicy: ClusterFirst
+        restartPolicy: Always
+        schedulerName: default-scheduler
+        securityContext: {}
+        terminationGracePeriodSeconds: 30
+        volumes:
+          - name: escerts
+            secret:
+              secretName: "log-anomaly-detector-odcs"
```
Comment (vpavlin) [inline openshift/official_template.yaml]: This seems like a very specific secret name mapped to a concrete (internal) use case/deployment - it should probably be templated or have a more general name

File: openshift/official_template.yaml [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+          - name: model-storage
+            persistentVolumeClaim:
+            claimName: log-anomaly-detector-f20
+          - configMap:
+              defaultMode: 420
+              name: "factory20-config"
```
Comment (vpavlin) [inline openshift/official_template.yaml]: Again, the "official template" hardcoding internal deployment details

**Comment by vpavlin** [comment]
Also - why is the name of the yaml file `official_template`? Do we have any unofficial templates? 

Why did we got rid of the `${NAME}` parameter in the template? 

I would also recommend to provide the error in the issue as it is not clear what exactly you are fixing - and why are you basically starting from scratch instead of correcting existing template:)

File: openshift/official_template.yaml (hunk 4 above)
Comment (zmhassan) [reply openshift/official_template.yaml]: So for each team that onboards this service and share the same namespace we should have unique names. In this case we use f20 its just a placeholder.

File: openshift/official_template.yaml (hunk 5 above)
Comment (zmhassan) [reply openshift/official_template.yaml]: Good point. I'll export this out as a PARAM

File: openshift/official_template.yaml (hunk 4 above)
Comment (vpavlin) [reply openshift/official_template.yaml]: Let me be more clear here - f20 should be a parameter (as it was in original template which you are removing in this PR) exaclty for a reason you mentioned - we need to have unique names for each deployed LAD

**Comment by zmhassan** [comment]
> Also - why is the name of the yaml file `official_template`? Do we have any unofficial templates?
> 
> Why did we got rid of the `${NAME}` parameter in the template?
> 
> I would also recommend to provide the error in the issue as it is not clear what exactly you are fixing - and why are you basically starting from scratch instead of correcting existing template:)

Hey, thank you for taking the time to review. I see that you probably have not run that old template. I'm going to provide a single template with that works. The previous template never worked for me.

File: openshift/official_template.yaml (hunk 1 above)
Comment (zmhassan) [reply openshift/official_template.yaml]: Fixed

File: openshift/official_template.yaml (hunk 2 above)
Comment (zmhassan) [reply openshift/official_template.yaml]: Fixed it

File: openshift/official_template.yaml (hunk 3 above)
Comment (zmhassan) [reply openshift/official_template.yaml]: Done

File: openshift/official_template.yaml (hunk 4 above)
Comment (zmhassan) [reply openshift/official_template.yaml]: Fixed it. Done.

File: openshift/official_template.yaml (hunk 6 above)
Comment (zmhassan) [reply openshift/official_template.yaml]: Fixed it. Done.

File: openshift/official_template.yaml (hunk 5 above)
Comment (zmhassan) [reply openshift/official_template.yaml]: Fixed it. Done.

**Comment by zmhassan** [comment]
Hey I've updated this PR and hope that you'll can review and merge if this looks good.

**Comment by zmhassan** [comment]
@durandom /all-check-pass /merge-if-it-looks-good

**Comment by zmhassan** [comment]
> please don't rename the yaml. The name is good and the README refers that filename
> 
> ```
> oc apply -f openshift/log-anomaly-detector.app.yaml
> ```
> 
> It's also hard to review the changes, because it looks like a new file.
> If you need to rename, do the rename in a different commit, so we can review the changes

Sorry. I have just fixed that and pushed another update. Please merge if this is good.

**Pull request merged**